core/drivers/aplic_priv.c

```
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021 Western Digital Corporation or its affiliates.
+ * Copyright (c) 2022 Ventana Micro Systems Inc.
+ * Copyright (c) 2025 Beijing Institute of Open Source Chip (BOSC)
+ *
+ * Authors:
+ *   Anup Patel <anup.patel@wdc.com>
+ *   Huang Borong <huangborong@bosc.ac.cn>
+ */
+
+#include <drivers/aplic_priv.h>
+#include <drivers/imsic.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <encoding.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <libfdt.h>
+#include <string.h>
+#include <tee_api_types.h>
+#include <trace.h>
+
+static TEE_Result fdt_parse_aplic_node(const void *fdt, int nodeoff,
+				       struct aplic_data *aplic)
+{
+	const fdt32_t *val = NULL;
+	struct imsic_data imsic = { };
+	int i = 0;
+	int len = 0;
+	int noff = -1;
+	int rc = -1;
+	paddr_t reg_addr = 0;
+	size_t reg_size = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (nodeoff < 0 || !aplic || !fdt)
+		return TEE_ERROR_GENERIC;
+	memset(aplic, 0, sizeof(*aplic));
+
+	rc = fdt_get_reg_props_by_index(fdt, nodeoff, 0, &reg_addr, &reg_size);
+	if (rc < 0 || !reg_addr || !reg_size)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	aplic->aplic_base = core_mmu_get_va(reg_addr, MEM_AREA_IO_SEC,
+					    reg_size);
+	if (!aplic->aplic_base)
+		return TEE_ERROR_GENERIC;
+	aplic->size = reg_size;
+
+	val = fdt_getprop(fdt, nodeoff, "riscv,num-sources", &len);
+	if (len > 0)
+		aplic->num_source = fdt32_to_cpu(*val);
+
+	val = fdt_getprop(fdt, nodeoff, "interrupts-extended", &len);
+	if (val && (size_t)len > sizeof(fdt32_t)) {
+		len = len / sizeof(fdt32_t);
+		for (i = 0; i < len; i += 2) {
+			if (fdt32_to_cpu(val[i + 1]) == IRQ_M_EXT) {
+				aplic->targets_mmode = true;
+				break;
+			}
+		}
+		aplic->num_idc = len / 2;
+	} else {
+		val = fdt_getprop(fdt, nodeoff, "msi-parent", &len);
+		if (val && (size_t)len >= sizeof(fdt32_t)) {
```

If there is no `interrupt-extended` nor `msi-parent` property, shouldn't this function return a error?
If `interrupt-extended` as only 1 cell, is it intentional to discard its value and consider only `msi-parent` property?

---

> If there is no `interrupt-extended` nor `msi-parent` property, shouldn't this function return a error?

I beleive that if the node has neither the `interrupt-extended` property nor the `msi-parent` property, it's an illegal node, and this function should return an error.

> If `interrupt-extended` as only 1 cell, is it intentional to discard its value and consider only msi-parent property?

If `interrupt-extended` has only 1 cell, it's also an illegal node, and this function should return an error.

Thank you for mentioning that. I will fix this in the next commit.